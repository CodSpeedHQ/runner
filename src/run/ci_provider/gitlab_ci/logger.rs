use console::style;
use lazy_static::lazy_static;
use log::{Level, LevelFilter, Log};
use regex::Regex;
use simplelog::SharedLogger;
use std::{
    env,
    io::Write,
    sync::Mutex,
    time::{SystemTime, UNIX_EPOCH},
};

use crate::{
    logger::{get_group_event, GroupEvent},
    run::ci_provider::logger::should_provider_logger_handle_record,
};

lazy_static! {
    static ref GITLAB_SECTION_ID_SANITIZE_REGEX: Regex =
        Regex::new(r"[^\d\w\-_]").expect("Failed to compile GitLab SectionId regex");
}

/// Unicode Escape character
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#general-ascii-codes
const U_ESC: char = '\x1B';

/// Unicode Carriage Return character
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#general-ascii-codes
const U_CR: char = '\x0D';

/// Reset color mode
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#colors--graphics-mode
const COLOR_RESET: &str = "\x1B[0m";

/// Erase cursor till end of line
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#erase-functions
const ERASE_CURSOR: &str = "\x1B[0K";

/// A logger that prints log in the format expected by GitLab CI
///
/// See https://docs.gitlab.com/ee/ci/yaml/script.html
pub struct GitLabCILogger {
    log_level: LevelFilter,
    section_id: Mutex<Option<String>>,
}

impl GitLabCILogger {
    pub fn new() -> Self {
        // force activation of colors, because GitlabCI does not
        // respect the CLICOLORS spec.
        // https://gitlab.com/gitlab-org/gitlab/-/issues/28598
        console::set_colors_enabled(true);

        let log_level = env::var("CODSPEED_LOG")
            .ok()
            .and_then(|log_level| log_level.parse::<log::LevelFilter>().ok())
            .unwrap_or(log::LevelFilter::Info);
        Self {
            log_level,
            section_id: Mutex::new(None),
        }
    }
}

impl Log for GitLabCILogger {
    fn enabled(&self, _metadata: &log::Metadata) -> bool {
        true
    }

    fn log(&self, record: &log::Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            let now = SystemTime::now();
            let timestamp = now.duration_since(UNIX_EPOCH).unwrap().as_secs();
            let mut section_id = self.section_id.lock().unwrap();

            match group_event {
                GroupEvent::Start(name) | GroupEvent::StartOpened(name) => {
                    let new_section_id = GITLAB_SECTION_ID_SANITIZE_REGEX
                        .replace_all(&name, "_")
                        .to_ascii_lowercase();

                    *section_id = Some(new_section_id.to_string());

                    // https://docs.gitlab.com/ee/ci/yaml/script.html#custom-collapsible-sections
                    println!("{ERASE_CURSOR}section_start:{timestamp}:{new_section_id}{U_CR}{ERASE_CURSOR}{U_ESC}[36;1m{name}{COLOR_RESET}");
                }
                GroupEvent::End => {
                    // do not fail if there is no current section
                    let current_section_id = section_id.clone().unwrap_or("".to_string());

                    // https://docs.gitlab.com/ee/ci/yaml/script.html#custom-collapsible-sections
                    println!("{ERASE_CURSOR}section_end:{timestamp}:{current_section_id}{U_CR}{ERASE_CURSOR}");

                    *section_id = None;
                }
            }
            return;
        }

        if level > self.log_level {
            return;
        }

        // set log colors. See https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#colors--graphics-mode
        match level {
            Level::Error => {
                println!("{}", style(message).red());
            }
            Level::Warn => {
                println!("{}", style(message).yellow());
            }
            Level::Info => {
                println!("{message}");
            }
            Level::Debug => {
                println!("{}", style(message).cyan());
            }
            Level::Trace => {
                println!("{}", style(message).magenta());
            }
        };
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for GitLabCILogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
