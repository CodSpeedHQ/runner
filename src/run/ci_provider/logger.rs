use crate::run::runner::VALGRIND_EXECUTION_TARGET;

pub(super) fn should_provider_logger_handle_record(record: &log::Record) -> bool {
    // Provider logger should handle all records except the ones from the valgrind execution target
    record.target() != VALGRIND_EXECUTION_TARGET
}
