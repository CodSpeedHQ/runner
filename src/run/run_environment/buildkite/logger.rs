use crate::{
    logger::{get_group_event, get_json_event, GroupEvent},
    run::run_environment::logger::should_provider_logger_handle_record,
};
use log::*;
use simplelog::SharedLogger;
use std::{env, io::Write};

/// A logger that prints logs in the format expected by Buildkite
///
/// See https://buildkite.com/docs/pipelines/managing-log-output
pub struct BuildkiteLogger {
    log_level: LevelFilter,
}

impl BuildkiteLogger {
    pub fn new() -> Self {
        let log_level = env::var("CODSPEED_LOG")
            .ok()
            .and_then(|log_level| log_level.parse::<log::LevelFilter>().ok())
            .unwrap_or(log::LevelFilter::Info);
        Self { log_level }
    }
}

impl Log for BuildkiteLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) => {
                    println!("--- {name}");
                }
                GroupEvent::StartOpened(name) => {
                    println!("+++ {name}");
                }
                GroupEvent::End => {}
            }
            return;
        }

        if get_json_event(record).is_some() {
            return;
        }

        if level > self.log_level {
            return;
        }
        // there is no support for log levels in Buildkite, so we print the level in the message
        match level {
            Level::Error => {
                println!("[ERROR] {message}");
            }
            Level::Warn => {
                println!("[WARN] {message}");
            }
            Level::Info => {
                println!("{message}");
            }
            Level::Debug => {
                println!("[DEBUG] {message}");
            }
            Level::Trace => {
                println!("[TRACE] {message}");
            }
        }
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for BuildkiteLogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
