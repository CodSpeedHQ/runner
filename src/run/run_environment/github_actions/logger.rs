use crate::{
    logger::{GroupEvent, get_group_event, get_json_event},
    run::run_environment::logger::should_provider_logger_handle_record,
};
use log::*;
use simplelog::SharedLogger;
use std::{env, io::Write};

/// A logger that prints logs in the format expected by GitHub Actions, with grouping support.
///
/// See https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions
pub struct GithubActionLogger {
    log_level: LevelFilter,
}

impl GithubActionLogger {
    pub fn new() -> Self {
        // Only enable debug logging if it's enabled in GitHub Actions.
        // See: https://docs.github.com/en/actions/reference/workflows-and-actions/variables
        let log_level = if env::var("RUNNER_DEBUG").unwrap_or_default() == "1" {
            LevelFilter::Trace
        } else {
            env::var("CODSPEED_LOG")
                .ok()
                .and_then(|log_level| log_level.parse::<LevelFilter>().ok())
                .unwrap_or(LevelFilter::Info)
        };

        Self { log_level }
    }
}

impl Log for GithubActionLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) | GroupEvent::StartOpened(name) => {
                    println!("::group::{name}");
                }
                GroupEvent::End => {
                    println!("::endgroup::");
                }
            }
            return;
        }

        if get_json_event(record).is_some() {
            return;
        }

        if level > self.log_level {
            return;
        }

        let prefix = match level {
            Level::Error => "::error::",
            Level::Warn => "::warning::",
            Level::Info => "",
            Level::Debug => "::debug::",
            Level::Trace => "::debug::[TRACE]",
        };
        let message_string = message.to_string();
        let lines = message_string.lines();
        // ensure that all the lines of the message have the prefix, otherwise GitHub Actions will not recognize the command for the whole string
        lines.for_each(|line| println!("{prefix}{line}"));
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for GithubActionLogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
