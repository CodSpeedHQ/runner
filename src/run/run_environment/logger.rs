use crate::run::runner::EXECUTOR_TARGET;

pub(super) fn should_provider_logger_handle_record(record: &log::Record) -> bool {
    // Provider logger should handle all records except the ones from the executor target
    record.target() != EXECUTOR_TARGET
}
