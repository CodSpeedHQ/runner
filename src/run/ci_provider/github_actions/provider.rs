use lazy_static::lazy_static;
use regex::Regex;
use serde_json::Value;
use simplelog::SharedLogger;
use std::{env, fs};

use crate::prelude::*;
use crate::run::{
    ci_provider::{
        interfaces::{GhData, ProviderMetadata, RunEvent, Sender},
        provider::{CIProvider, CIProviderDetector},
    },
    config::Config,
    helpers::{find_repository_root, get_env_variable},
};

use super::logger::GithubActionLogger;

#[derive(Debug)]
pub struct GitHubActionsProvider {
    pub owner: String,
    pub repository: String,
    pub ref_: String,
    pub head_ref: Option<String>,
    pub base_ref: Option<String>,
    pub gh_data: GhData,
    pub event: RunEvent,
    pub repository_root_path: String,
}

impl GitHubActionsProvider {
    fn get_owner_and_repository() -> Result<(String, String)> {
        let owner_and_repository = get_env_variable("GITHUB_REPOSITORY")?;
        let mut owner_and_repository = owner_and_repository.split('/');
        let owner = owner_and_repository.next().unwrap();
        let repository = owner_and_repository.next().unwrap();
        Ok((owner.into(), repository.into()))
    }
}

lazy_static! {
    static ref PR_REF_REGEX: Regex = Regex::new(r"^refs/pull/(?P<pr_number>\d+)/merge$").unwrap();
}

impl TryFrom<&Config> for GitHubActionsProvider {
    type Error = Error;
    fn try_from(_config: &Config) -> Result<Self> {
        let (owner, repository) = Self::get_owner_and_repository()?;
        let ref_ = get_env_variable("GITHUB_REF")?;
        let is_pr = PR_REF_REGEX.is_match(&ref_);
        let head_ref = if is_pr {
            let github_event_path = get_env_variable("GITHUB_EVENT_PATH")?;
            let github_event = fs::read_to_string(github_event_path)?;
            let github_event: Value = serde_json::from_str(&github_event)
                .expect("GITHUB_EVENT_PATH file could not be read");
            let pull_request = github_event["pull_request"].as_object().unwrap();

            let head_repo = pull_request["head"]["repo"].as_object().unwrap();
            let base_repo = pull_request["base"]["repo"].as_object().unwrap();

            let is_head_repo_fork = head_repo["id"] != base_repo["id"];

            let head_ref = if is_head_repo_fork {
                format!(
                    "{}:{}",
                    head_repo["owner"]["login"].as_str().unwrap(),
                    pull_request["head"]["ref"].as_str().unwrap()
                )
            } else {
                pull_request["head"]["ref"].as_str().unwrap().to_owned()
            };
            Some(head_ref)
        } else {
            None
        };

        let github_event_name = get_env_variable("GITHUB_EVENT_NAME")?;
        let event = serde_json::from_str(&format!("\"{}\"", github_event_name)).context(
            format!("Event {} is not supported by CodSpeed", github_event_name),
        )?;
        let repository_root_path = match find_repository_root(&std::env::current_dir()?) {
            Some(mut path) => {
                // Add a trailing slash to the path
                path.push("");
                path.to_string_lossy().to_string()
            }
            None => format!("/home/runner/work/{}/{}/", repository, repository),
        };

        Ok(Self {
            owner,
            repository: repository.clone(),
            ref_,
            head_ref,
            event,
            gh_data: GhData {
                job: get_env_variable("GITHUB_JOB")?,
                run_id: get_env_variable("GITHUB_RUN_ID")?
                    .parse()
                    .context("Failed to parse GITHUB_RUN_ID into an integer")?,
                sender: Some(Sender {
                    login: get_env_variable("GITHUB_ACTOR")?,
                    id: get_env_variable("GITHUB_ACTOR_ID")?
                        .parse()
                        .context("Failed to parse GITHUB_ACTOR_ID into an integer")?,
                }),
            },
            base_ref: get_env_variable("GITHUB_BASE_REF").ok(),
            repository_root_path,
        })
    }
}

impl CIProviderDetector for GitHubActionsProvider {
    fn detect() -> bool {
        // check if the GITHUB_ACTIONS environment variable is set and the value is truthy
        env::var("GITHUB_ACTIONS") == Ok("true".into())
    }
}

impl CIProvider for GitHubActionsProvider {
    fn get_logger(&self) -> Box<dyn SharedLogger> {
        Box::new(GithubActionLogger)
    }

    fn get_provider_name(&self) -> &'static str {
        "GitHub Actions"
    }

    fn get_provider_slug(&self) -> &'static str {
        "github-actions"
    }

    fn get_provider_metadata(&self) -> Result<ProviderMetadata> {
        Ok(ProviderMetadata {
            base_ref: self.base_ref.clone(),
            head_ref: self.head_ref.clone(),
            event: self.event.clone(),
            gh_data: Some(self.gh_data.clone()),
            owner: self.owner.clone(),
            repository: self.repository.clone(),
            ref_: self.ref_.clone(),
            repository_root_path: self.repository_root_path.clone(),
        })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;
    use temp_env::{with_var, with_vars};

    use crate::VERSION;

    use super::*;

    #[test]
    fn test_detect() {
        with_var("GITHUB_ACTIONS", Some("true"), || {
            assert!(GitHubActionsProvider::detect());
        });
    }

    #[test]
    fn test_get_owner_and_repository() {
        with_var("GITHUB_REPOSITORY", Some("owner/repository"), || {
            let (owner, repository) = GitHubActionsProvider::get_owner_and_repository().unwrap();
            assert_eq!(owner, "owner");
            assert_eq!(repository, "repository");
        });
    }

    #[test]
    fn test_try_from_push_main() {
        with_vars(
            [
                ("GITHUB_ACTOR_ID", Some("1234567890")),
                ("GITHUB_ACTOR", Some("actor")),
                ("GITHUB_BASE_REF", Some("main")),
                ("GITHUB_EVENT_NAME", Some("push")),
                ("GITHUB_JOB", Some("job")),
                ("GITHUB_REF", Some("refs/heads/main")),
                ("GITHUB_REPOSITORY", Some("owner/repository")),
                ("GITHUB_RUN_ID", Some("1234567890")),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let github_actions_provider = GitHubActionsProvider::try_from(&config).unwrap();
                assert_eq!(github_actions_provider.owner, "owner");
                assert_eq!(github_actions_provider.repository, "repository");
                assert_eq!(github_actions_provider.ref_, "refs/heads/main");
                assert_eq!(github_actions_provider.base_ref, Some("main".into()));
                assert_eq!(github_actions_provider.head_ref, None);
                assert_eq!(github_actions_provider.event, RunEvent::Push);
                assert_eq!(github_actions_provider.gh_data.job, "job");
                assert_eq!(github_actions_provider.gh_data.run_id, 1234567890);
                assert_eq!(
                    github_actions_provider
                        .gh_data
                        .sender
                        .as_ref()
                        .unwrap()
                        .login,
                    "actor"
                );
                assert_eq!(
                    github_actions_provider.gh_data.sender.as_ref().unwrap().id,
                    1234567890
                );
            },
        )
    }

    #[test]
    fn test_pull_request_provider_metadata() {
        with_vars(
            [
                ("GITHUB_ACTIONS", Some("true")),
                ("GITHUB_ACTOR_ID", Some("19605940")),
                ("GITHUB_ACTOR", Some("adriencaccia")),
                ("GITHUB_BASE_REF", Some("main")),
                ("GITHUB_EVENT_NAME", Some("pull_request")),
                (
                    "GITHUB_EVENT_PATH",
                    Some(
                        format!(
                            "{}/src/run/ci_provider/github_actions/samples/pr-event.json",
                            env!("CARGO_MANIFEST_DIR")
                        )
                        .as_str(),
                    ),
                ),
                ("GITHUB_HEAD_REF", Some("feat/codspeed-runner")),
                ("GITHUB_JOB", Some("log-env")),
                ("GITHUB_REF", Some("refs/pull/22/merge")),
                ("GITHUB_REPOSITORY", Some("my-org/adrien-python-test")),
                ("GITHUB_RUN_ID", Some("6957110437")),
                ("VERSION", Some("0.1.0")),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let github_actions_provider = GitHubActionsProvider::try_from(&config).unwrap();
                let provider_metadata = github_actions_provider.get_provider_metadata().unwrap();

                assert_json_snapshot!(provider_metadata, {
                    ".runner.version" => insta::dynamic_redaction(|value,_path| {
                        assert_eq!(value.as_str().unwrap(), VERSION.to_string());
                        "[version]"
                    }),
                });
            },
        );
    }

    #[test]
    fn test_fork_pull_request_provider_metadata() {
        with_vars(
            [
                ("GITHUB_ACTIONS", Some("true")),
                ("GITHUB_ACTOR_ID", Some("19605940")),
                ("GITHUB_ACTOR", Some("adriencaccia")),
                ("GITHUB_BASE_REF", Some("main")),
                ("GITHUB_EVENT_NAME", Some("pull_request")),
                (
                    "GITHUB_EVENT_PATH",
                    Some(
                        format!(
                            "{}/src/run/ci_provider/github_actions/samples/fork-pr-event.json",
                            env!("CARGO_MANIFEST_DIR")
                        )
                        .as_str(),
                    ),
                ),
                ("GITHUB_HEAD_REF", Some("feat/codspeed-runner")),
                ("GITHUB_JOB", Some("log-env")),
                ("GITHUB_REF", Some("refs/pull/22/merge")),
                ("GITHUB_REPOSITORY", Some("my-org/adrien-python-test")),
                ("GITHUB_RUN_ID", Some("6957110437")),
                ("VERSION", Some("0.1.0")),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let github_actions_provider = GitHubActionsProvider::try_from(&config).unwrap();
                let provider_metadata = github_actions_provider.get_provider_metadata().unwrap();

                assert_eq!(provider_metadata.owner, "my-org");
                assert_eq!(provider_metadata.repository, "adrien-python-test");
                assert_eq!(provider_metadata.base_ref, Some("main".into()));
                assert_eq!(
                    provider_metadata.head_ref,
                    Some("fork-owner:feat/codspeed-runner".into())
                );
                assert_json_snapshot!(provider_metadata, {
                    ".runner.version" => insta::dynamic_redaction(|value,_path| {
                        assert_eq!(value.as_str().unwrap(), VERSION.to_string());
                        "[version]"
                    }),
                });
            },
        );
    }
}
