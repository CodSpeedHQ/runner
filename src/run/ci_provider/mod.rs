pub mod interfaces;
pub mod logger;
mod provider;

use buildkite::BuildkiteProvider;
use github_actions::GitHubActionsProvider;
use local::LocalProvider;
use provider::CIProviderDetector;

use crate::prelude::*;
use crate::run::config::Config;

pub use self::provider::CIProvider;

// Provider implementations
mod buildkite;
mod github_actions;
mod local;

pub fn get_provider(config: &Config) -> Result<Box<dyn CIProvider>> {
    if BuildkiteProvider::detect() {
        let provider = BuildkiteProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if GitHubActionsProvider::detect() {
        let provider = GitHubActionsProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if LocalProvider::detect() {
        let provider = LocalProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    bail!("No CI provider detected")
}
