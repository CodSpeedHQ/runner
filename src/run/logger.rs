use crate::logger::{GROUP_TARGET, OPENED_GROUP_TARGET};
use crate::prelude::*;
use crate::run::{ci_provider::CIProvider, runner::RunData};
use log::LevelFilter;
use simplelog::{CombinedLogger, WriteLogger};
use std::fs::copy;
use std::path::PathBuf;
use tempfile::NamedTempFile;

pub struct Logger {
    log_file_path: PathBuf,
}

impl Logger {
    #[allow(clippy::borrowed_box)]
    pub fn new(provider: &Box<dyn CIProvider>) -> Result<Self> {
        let provider_logger = provider.get_logger();
        let log_file = NamedTempFile::new().context("Failed to create log file")?;
        let log_file_path = log_file.path().to_path_buf();
        let file_logger_config = simplelog::ConfigBuilder::new()
            // Groups are not logged to the file
            .add_filter_ignore_str(GROUP_TARGET)
            .add_filter_ignore_str(OPENED_GROUP_TARGET)
            .build();
        let file_logger = WriteLogger::new(LevelFilter::Trace, file_logger_config, log_file);
        CombinedLogger::init(vec![provider_logger, file_logger])
            .context("Failed to init logger")?;
        Ok(Self { log_file_path })
    }

    pub fn persist_log_to_profile_folder(&self, run_data: &RunData) -> Result<()> {
        let profile_folder = run_data.profile_folder.clone();
        let dest_log_file_path = profile_folder.join("runner.log");
        debug!("Persisting log file to {}", dest_log_file_path.display());
        log::logger().flush();
        copy(&self.log_file_path, dest_log_file_path).context("Failed to copy log file")?;
        Ok(())
    }
}
