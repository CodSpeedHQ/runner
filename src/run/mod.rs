use crate::api_client::CodSpeedAPIClient;
use crate::config::CodSpeedConfig;
use crate::prelude::*;
use crate::run::{config::Config, logger::Logger};
use crate::VERSION;
use check_system::SystemInfo;
use clap::Args;
use instruments::mongo_tracer::MongoTracer;
use runner::get_run_data;

mod check_system;
pub mod ci_provider;
mod helpers;
mod instruments;
mod poll_results;
mod runner;
mod uploader;

pub mod config;
pub mod logger;

fn show_banner() {
    let banner = format!(
        r#"
   ______            __ _____                         __
  / ____/____   ____/ // ___/ ____   ___   ___   ____/ /
 / /    / __ \ / __  / \__ \ / __ \ / _ \ / _ \ / __  /
/ /___ / /_/ // /_/ / ___/ // /_/ //  __//  __// /_/ /
\____/ \____/ \__,_/ /____// .___/ \___/ \___/ \__,_/
  https://codspeed.io     /_/          runner v{}
"#,
        VERSION
    );
    println!("{}", banner);
    debug!("codspeed v{}", VERSION);
}

#[derive(Args, Debug)]
pub struct RunArgs {
    /// The upload URL to use for uploading the results, useful for on-premises installations
    #[arg(long)]
    pub upload_url: Option<String>,

    /// The token to use for uploading the results,
    #[arg(long, env = "CODSPEED_TOKEN")]
    pub token: Option<String>,

    /// The directory where the command will be executed.
    #[arg(long)]
    pub working_directory: Option<String>,

    /// Comma-separated list of instruments to enable. Possible values: mongodb.
    #[arg(long, value_delimiter = ',')]
    pub instruments: Vec<String>,

    /// The name of the environment variable that contains the MongoDB URI to patch.
    /// If not provided, user will have to provide it dynamically through a CodSpeed integration.
    ///
    /// Only used if the `mongodb` instrument is enabled.
    #[arg(long)]
    pub mongo_uri_env_name: Option<String>,

    /// Only for debugging purposes, skips the upload of the results
    #[arg(
        long,
        default_value = "false",
        hide = true,
        env = "CODSPEED_SKIP_UPLOAD"
    )]
    pub skip_upload: bool,

    /// Only for debugging purposes, skips the setup of the runner
    #[arg(long, default_value = "false", hide = true)]
    pub skip_setup: bool,

    /// The bench command to run
    pub command: Vec<String>,
}

#[cfg(test)]
impl RunArgs {
    /// Constructs a new `RunArgs` with default values for testing purposes
    pub fn test() -> Self {
        Self {
            upload_url: None,
            token: None,
            working_directory: None,
            instruments: vec![],
            mongo_uri_env_name: None,
            skip_upload: false,
            skip_setup: false,
            command: vec![],
        }
    }
}

pub async fn run(args: RunArgs, api_client: &CodSpeedAPIClient) -> Result<()> {
    let mut config = Config::try_from(args)?;
    let provider = ci_provider::get_provider(&config)?;
    let codspeed_config = CodSpeedConfig::load()?;
    let logger = Logger::new(&provider)?;

    if provider.get_provider_slug() != "local" {
        show_banner();
    }
    debug!("config: {:#?}", config);

    if provider.get_provider_slug() == "local" {
        if codspeed_config.auth.token.is_none() {
            bail!("You have to authenticate the CLI first. Run `codspeed auth login`.");
        }
        debug!("Using the token from the CodSpeed configuration file");
        config.set_token(codspeed_config.auth.token.clone());
    }

    let system_info = SystemInfo::new()?;
    check_system::check_system(&system_info)?;

    let mode = runner::get_mode()?;
    let executor = runner::get_executor_from_mode(mode);

    let run_data = get_run_data()?;

    if !config.skip_setup {
        start_group!("Preparing the environment");
        executor.setup(&config, &system_info, &run_data).await?;
        end_group!();
    }

    start_opened_group!("Running the benchmarks");

    // TODO: refactor and move directly in the Instruments struct as a `start` method
    let mongo_tracer = if let Some(mongodb_config) = &config.instruments.mongodb {
        let mut mongo_tracer = MongoTracer::try_from(&run_data.profile_folder, mongodb_config)?;
        mongo_tracer.start().await?;
        Some(mongo_tracer)
    } else {
        None
    };

    executor
        .run(&config, &system_info, &run_data, &mongo_tracer)
        .await?;

    // TODO: refactor and move directly in the Instruments struct as a `stop` method
    if let Some(mut mongo_tracer) = mongo_tracer {
        mongo_tracer.stop().await?;
    }

    executor.teardown(&config, &system_info, &run_data).await?;

    end_group!();

    if !config.skip_upload {
        start_group!("Uploading performance data");
        logger.persist_log_to_profile_folder(&run_data)?;
        let upload_result =
            uploader::upload(&config, &system_info, &provider, &run_data, executor.name()).await?;
        end_group!();

        if provider.get_provider_slug() == "local" {
            start_group!("Fetching the results");
            poll_results::poll_results(api_client, &provider, upload_result.run_id).await?;
            end_group!();
        }
    }

    Ok(())
}
