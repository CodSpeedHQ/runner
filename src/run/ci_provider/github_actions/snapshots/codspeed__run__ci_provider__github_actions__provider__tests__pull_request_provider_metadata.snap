---
source: src/run/ci_provider/github_actions/provider.rs
expression: provider_metadata
---
{
  "ref": "refs/pull/22/merge",
  "headRef": "feat/codspeed-runner",
  "baseRef": "main",
  "owner": "my-org",
  "repository": "adrien-python-test",
  "event": "pull_request",
  "sender": {
    "id": "19605940",
    "login": "adriencaccia"
  },
  "ghData": {
    "runId": "6957110437",
    "job": "log-env"
  },
  "repositoryRootPath": "/home/runner/work/adrien-python-test/adrien-python-test/"
}
